GoogleForCreators/web-stories-wp
# Design System: Add Unit + Integer variants to input

## Context

Building out the design system. These updates for the input are necessary in the editor.

## Summary

Adds the `Unit` and `Numeric` input variants as described in #6414 and this comment: https://github.com/google/web-stories-wp/pull/6355#issuecomment-780749936

## Relevant Technical Choices

The `Input` component is meant to be a 'presentational' component. However, the functionality of 'reverting to a previous value' makes this component more of a 'container' component.

I am keeping the `Input` as presentational as possible, but wrapping it in a container component to add functionality that the 'numeric' input will need (saving old inputs, only calling `onChange` when the value should be "submitted", etc.). Other changes:
- `allowNegative` is redundant when paired with `max` and `min`. Removed it so that we use only max/min. If there is a bound, we can specify it there.
- I kept the API similar to the old input rather than using `type="numeric"` for the input. I would prefer to use `type="numeric"` but there was already old logic written that I used instead of making a brand new component
- created this without setting event listeners on the window. This should be more performant and relies on only "react"
- added a hook to manage the `Numeric` input's state. This also makes testing easier.

## To-do

After this PR is merged I'd like to make one final PR that merges `Input` and `Numeric` into one component. That component will decide which input type to use based on a `type` prop.
- if `type="numeric"` - render the `<Numeric />` input
- else render the base `<Input />`

## User-facing changes



none

## Testing Instructions



Visit storybook. Design System -> Input -> Numeric

Test the following scenarios:
- can enter any value while typing
- blurring the field will trigger validation. If there is a valid number, the number will remain.
    - if the value is invalid and the previous value was valid, leave the valid number
    - if the value is invalid and the previous value was also invalid, leave a new valid number
- `isFloat=true` will allow for float values as well as integers
- arrow keys: ⬆️  and ⬇️  changes the value 1 up or down
- arrow keys + alt: ⬆️  and ⬇️  changes the value by 0.1 up or down (if `isFloat` is `true`)
- `unit` should only be visible when the field is not focused
- field cannot be empty unless `allowEmpty` is `true`

All of these values can be configured in the `knobs` section in storybook

### QA






### UAT







This PR can be tested by following these steps:


Visit storybook. Design System -> Input -> Numeric

Test the following scenarios:
- can enter any value while typing
- blurring the field will trigger validation. If there is a valid number, the number will remain.
    - if the value is invalid and the previous value was valid, leave the valid number
    - if the value is invalid and the previous value was also invalid, leave a new valid number
- `isFloat=true` will allow for float values as well as integers
- arrow keys: ⬆️  and ⬇️  changes the value 1 up or down
- arrow keys + alt: ⬆️  and ⬇️  changes the value by 0.1 up or down (if `isFloat` is `true`)
- `unit` should only be visible when the field is not focused
- field cannot be empty unless `allowEmpty` is `true`

All of these values can be configured in the `knobs` section in storybook


|||
|--|--|
|Numeric input|<image src="https://user-images.githubusercontent.com/22185279/108396951-aaf32900-71d4-11eb-9c39-6e30de3dee11.gif" width="400px" />|
|Unit|<image src="https://user-images.githubusercontent.com/22185279/108396946-a9296580-71d4-11eb-8528-a2480895e020.gif" width="400px" />|
|Minimum|<image src="https://user-images.githubusercontent.com/22185279/108396933-a62e7500-71d4-11eb-8850-61442f15dd20.gif" width="200px" />|
|Maximum|<image src="https://user-images.githubusercontent.com/22185279/108396939-a890cf00-71d4-11eb-8d02-fc4362078cbf.gif" width="200px" />|
|`ArrowUp` and `ArrowDown`|<image src="https://user-images.githubusercontent.com/22185279/108396947-a9c1fc00-71d4-11eb-8ba7-62d40420f3eb.gif" width="400px" />|
|`Esc` key press|<image src="https://user-images.githubusercontent.com/22185279/108396949-aa5a9280-71d4-11eb-9a11-d9e2ac5df671.gif" width="400px" />|

## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #6414
